ledger/participant-state/kvutils/tools/entries-extractor/src/replay/scala/ledger/participant/state/kvutils/tools/extractor/Extractor.scala

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils.tools.extractor
+
+import com.daml.ledger.participant.state.kvutils.Conversions._
+import com.daml.ledger.participant.state.kvutils.export.{
+  ProtobufBasedLedgerDataImporter,
+  SubmissionInfo,
+}
+import com.daml.ledger.participant.state.kvutils.wire.DamlSubmission
+import com.daml.ledger.participant.state.kvutils.{Envelope, Raw}
+import com.daml.lf.data._
+import com.daml.lf.testing.snapshot.Snapshot
+
+import java.io.BufferedOutputStream
+import java.nio.file.{Files, Path, Paths}
+import scala.jdk.CollectionConverters._
+import scala.util.control.NonFatal
+
+object Extractor extends App {
+
+  private[this] def decodeSubmission(
+      participantId: Ref.ParticipantId,
+      submission: DamlSubmission,
+  ): LazyList[Snapshot.SubmissionEntry] = {
+    submission.getPayloadCase match {
+      case DamlSubmission.PayloadCase.TRANSACTION_ENTRY =>
+        val entry = submission.getTransactionEntry
+        LazyList(
+          Snapshot.SubmissionEntry
+            .newBuilder()
+            .setTransaction(
+              Snapshot.TransactionEntry
+                .newBuilder()
+                .setRawTransaction(entry.getRawTransaction)
+                .setParticipantId(participantId)
+                .addAllSubmitters(entry.getSubmitterInfo.getSubmittersList)
+                .setLedgerTime(parseTimestamp(entry.getLedgerEffectiveTime).micros)
+                .setSubmissionTime(parseTimestamp(entry.getSubmissionTime).micros)
+                .setSubmissionSeed(entry.getSubmissionSeed)
+            )
+            .build()
+        )
+      case DamlSubmission.PayloadCase.PACKAGE_UPLOAD_ENTRY =>
+        val entry = submission.getPackageUploadEntry
+        entry.getArchivesList.asScala.iterator
+          .map(
+            Snapshot.SubmissionEntry
+              .newBuilder()
+              .setArchives(_)
+              .build()
+          )
+          .to(LazyList)
+      case _ =>
+        LazyList.empty
+    }
+  }
+
+  private[this] def decodeEnvelope(
+      participantId: Ref.ParticipantId,
+      envelope: Raw.Envelope,
+  ): LazyList[Snapshot.SubmissionEntry] =
+    assertRight(Envelope.open(envelope)) match {
+      case Envelope.SubmissionMessage(submission) =>
+        decodeSubmission(participantId, submission)
+      case Envelope.SubmissionBatchMessage(batch) =>
+        batch.getSubmissionsList.asScala
+          .to(LazyList)
+          .map(_.getSubmission)
+          .flatMap(submissionEnvelope =>
+            decodeEnvelope(participantId, Raw.Envelope(submissionEnvelope))
+          )
+      case Envelope.LogEntryMessage(_) | Envelope.StateValueMessage(_) =>
+        LazyList.empty
+    }
+
+  private[this] def decodeSubmissionInfo(submissionInfo: SubmissionInfo) =
```

Comment:
I think the methods should be in the opposite order to follow the execution order. This method, for instance, is the first one, as it's being run by the app/main.

Reply:
:+1: